xarray/core/dtypes.py

```diff
@@ -146,24 +147,83 @@ def get_neg_infinity(dtype, min_for_int=False):
     -------
     fill_value : positive infinity value corresponding to this dtype.
     """
-    if issubclass(dtype.type, np.floating):
+    if isdtype(dtype, "real floating"):
         return -np.inf
 
-    if issubclass(dtype.type, np.integer):
+    if isdtype(dtype, "integral"):
         if min_for_int:
             return np.iinfo(dtype).min
         else:
             return -np.inf
 
-    if issubclass(dtype.type, np.complexfloating):
+    if isdtype(dtype, "complex floating"):
         return -np.inf - 1j * np.inf
 
     return NINF
 
 
 def is_datetime_like(dtype):
     """Check if a dtype is a subclass of the numpy datetime types"""
-    return np.issubdtype(dtype, np.datetime64) or np.issubdtype(dtype, np.timedelta64)
+    return isdtype(dtype, (np.datetime64, np.timedelta64))
+
+
+def isdtype(dtype, kind, xp=None):
+    array_api_names = {
+        "bool": np.bool_,
+        "signed integer": np.signedinteger,
+        "unsigned integer": np.unsignedinteger,
+        "integral": np.integer,
+        "real floating": np.floating,
+        "complex floating": np.complexfloating,
+        "numeric": np.number,
+    }
+    numpy_names = {
+        "object": np.object_,
+        "character": np.character,
+        "string": np.str_,
+    }
+    long_names = array_api_names | numpy_names
+
+    def is_numpy_kind(kind):
+        return (isinstance(kind, str) and kind in numpy_names) or (
+            isinstance(kind, type) and issubclass(kind, (np.dtype, np.generic))
+        )
+
+    def split_numpy_kinds(kinds):
+        if not isinstance(kinds, tuple):
+            kinds = (kinds,)
+
+        numpy_kinds = tuple(kind for kind in kinds if is_numpy_kind(kind))
+        non_numpy_kinds = tuple(kind for kind in kinds if not is_numpy_kind(kind))
+
+        return numpy_kinds, non_numpy_kinds
+
+    def numpy_isdtype(dtype, kinds):
+        translated_kinds = [long_names.get(kind, kind) for kind in kinds]
+        if isinstance(dtype, np.generic):
+            return any(isinstance(dtype, kind) for kind in translated_kinds)
+        else:
+            return any(np.issubdtype(dtype, kind) for kind in translated_kinds)
+
+    if xp is None:
+        xp = np
+
+    if not isinstance(kind, tuple):
+        kinds = (kind,)
+    else:
+        kinds = kind
+
+    if isinstance(dtype, np.dtype):
+        return numpy_isdtype(dtype, kinds)
+    elif is_extension_array_dtype(dtype):
+        return any(dtype == kind for kind in kinds)
+    else:
+        numpy_kinds, non_numpy_kinds = split_numpy_kinds(kind)
+
+        if not non_numpy_kinds:
+            return False
```

Comment:
this assumes that we only get `numpy` dtypes, `pandas` extension array dtypes, or dtypes from a Array API compliant library. If we should expect other dtypes (I certainly can't anything), we'd have to raise an error if `xp` is `None` (and remove the fallback to `numpy` above)